server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -173,10 +169,41 @@ public DataStream(
         List<Index> failureIndices,
         boolean rolloverOnWrite,
         @Nullable DataStreamAutoShardingEvent autoShardingEvent
+    ) {
+        this(
+            name,
+            generation,
+            metadata,
+            hidden,
+            replicated,
+            system,
+            timeProvider,
+            allowCustomRouting,
+            indexMode,
+            lifecycle,
+            failureStoreEnabled,
+            new DataStreamIndices(BACKING_INDEX_PREFIX, List.copyOf(indices), rolloverOnWrite, autoShardingEvent),
+            new DataStreamIndices(FAILURE_STORE_PREFIX, failureStoreEnabled ? List.copyOf(failureIndices) : List.of(), false, null)
+        );
+    }
```

Comment:
I don't mind the idea of keeping what you have in the record class for making specific builders for backing/failure indices and just updating those test classes. This constructor was already as internal as it could be, so it probably doesn't need to stick around if there's a better way to do what it needed to do.

Reply:
I'm not sure I completely get what you're suggesting. Are you suggesting to just use the `DataStreamIndices` builder(s) I currently have in those `Test` usages? And not to add constructor aliases like the one I wrote in my first comment in this thread?

Reply:
Okay so turns out N - 1 of those usages I mentioned before were specifying `System::currentTimeMillis` as `timeProvider`, so I could just change them to use the public/main constructor. `DataStreamTestHelper#randomInstance` was the only usage that was specifying a custom `timeProvider`, so I made use of the `DataStreamIndices` there.